Algorithm.Framework/Portfolio/MaximumSharpeRatioPortfolioOptimizer.cs

```diff
@@ -0,0 +1,48 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using Accord.Math;
+using Accord.Math.Optimization;
+using Accord.Statistics;
+
+namespace QuantConnect.Algorithm.Framework.Portfolio
+{
+    public class MaximumSharpeRatioPortfolioOptimizer : MinimumVariancePortfolioOptimizer
+    {
+        private double _riskFreeRate;
+
+        public MaximumSharpeRatioPortfolioOptimizer(double lower = -1, double upper = 1, double riskFreeRate = 0.0) : base(lower, upper, 0.0)
+        {
+            _riskFreeRate = riskFreeRate;
+        }
+
+        public override double[] Optimize(double[,] historicalReturns, double[] expectedReturns = null)
+        {
+            _cov = historicalReturns.Covariance();
+
+            // (r -r_f)^T x = R or (r -r_f)^T x >= 0
+            _constraints.Add(new LinearConstraint(Size)
+            {
+                CombinedAs = (expectedReturns ?? historicalReturns.Mean(0)).Subtract(_riskFreeRate),
+                ShouldBe = _targetReturn > 0 ? ConstraintType.EqualTo : ConstraintType.GreaterThanOrEqualTo,
+                Value = _targetReturn
+            });
```

Comment:
Should it not be `expectedReturns` instead of `historicalReturns.Mean(0))`? 

It looks like you are setting a constraint that the portfolio excess return should be greater than 0 while co-variance is minimized. Does it ensure maximum Sharpe Ratio?

If it is possible, I would rather see a function/method that computes Sharpe Ratio so that we could print out the `solver` output and see the portfolio Sharpe Ratio.

Reply:
>Should it not be `expectedReturns` instead of `historicalReturns.Mean(0)`?

It is `expectedReturns` unless it is `null` which is allowed by method signature.

> It looks like you are setting a constraint that the portfolio excess return should be greater than 0 while co-variance is minimized. Does it ensure maximum Sharpe Ratio?

You're right. I missed this part. Constraint should be following:
```
// (r -r_f)^T x = 1
_constraints.Add(new LinearConstraint(Size)
{
    CombinedAs = (expectedReturns ?? historicalReturns.Mean(0)).Subtract(_riskFreeRate),
    ShouldBe = ConstraintType.EqualTo,
    Value = 1.0
 });
```

This is common approximation of the max SR to quadratic program. See: https://quant.stackexchange.com/questions/18521/sharpe-maximization-under-quadratic-constraints

> If it is possible, I would rather see a function/method that computes Sharpe Ratio so that we could print out the solver output and see the portfolio Sharpe Ratio.

No, it's not possible. The only Accord solver that supports linear constraints is QP solver.

Reply:
BTW, Accord's Cobyla solver hangs if boundary constrains are set, and L-BFGS isn't usable as it requires the gradient of the optimized function.

Reply:
You should assign `expectedReturns = expectedReturns ?? historicalReturns.Mean(0)` before adding to the constraints, because otherwise `.Mean(0)` will be called on every iteration to yield the same result. Better yet, you could create a new variable to include the `_riskFreeRate`:
```csharp
var excessExpectedReturn = (expectedReturns ?? historicalReturns.Mean(0)).Subtract(_riskFreeRate);
```

Reply:
I see.